### ISSUE TITLE
No socket creation during request

### ISSUE THREAD

**Issue by lolngh** [libwww-perl/libwww-perl#458]
Hi team,

Hope you are doing well. We have NTLM Sharepoint application configured in our application. The library we are using to support the authentication operation/feature is **LWP::Authen::Ntlm and Net::SSL.** The authentication flow is similar to what is posted in the [page’s](https://metacpan.org/pod/LWP::Authen::Ntlm) synopsis. It was working fine with **apache2-2.4.51(without tls1.3) and openssl-1.0.2p**.

But recently we upgraded the apache2 library to support **tls1.3(with openssl-1.1.1d)**. After this upgrade, the app is not able to make a socket connection to the sharepoint server. When debugged further - it wasn't able to create an ssl-socket using **Crypt::SSLeay::Conn->new in Net::SSL(v2.85)** module.

We are able to connect to the sharepoint server through openssl command(openssl s_client connect server:port).
The request made via the LWP module to the sharepoint-server is through **TLS-1.2** protocol.

Could you please let us know what might be the possible reasons for this issue and what are the possible ways to resolve this issue?

**Perl version being used: perl-5.18.2
Apache2 mod-perl : apache2-mod_perl-2.0.8-13.5.1.x86_64
LWP based modules : perl-libwww-perl-6.05, perl-LWP-MediaTypes-6.02, perl-LWP-Protocol-https-6.04, perl-IO-Socket-SSL-2.052-4.3.1, perl-Crypt-SSLeay-0.64-3.57, perl-NTLM-1.09-1.1, perl-Net-HTTP-6.06-5.2, perl-Net-SSLeay-1.64-3.2.**

**Comment by oalders** [comment]
Perhaps @noxxi has thoughts on this.

**Comment by noxxi** [comment]
Net::SSL + Crypt::SSLeay is not maintained anymore for 10 years. Please use IO::Socket::SSL + Net::SSLeay instead, which is the default for LWP::Protocol::https unless explicitly overwritten. Please also use recent versions of LWP, IO::Socket::SSL, Net::SSLeay etc, the versions you use here also also many years old and unsupported. Don't expect new features like TLS 1.3 to work with terribly old versions. For example support for TLS 1.3 was only added with IO::Socket::SSL 2.060 (you are using 2.052).

**Comment by lolngh** [comment]
Hi @noxxi ,

Will give a try and get back to you.

**Comment by noxxi** [comment]
If you really get a 500 from the server it is not a TLS problem anymore since the TLS handshake was successful - otherwise the  HTTP error could not have been sent. I'm not sure though that the 500 is really generated by the server or a pseudo error generated by LWP. 

To debug this I recommend that you have a closer look at the error status message and content, not only at the status code. As for debugging SSL you can run your code with `perl -MIO::Socket::SSL=debug9 app.pl` to observe what is going on at the SSL level.

**Comment by lolngh** [comment]
The content of resposne is **Can't connect to <server>:<port>**
Let me try to debug through SSL using above command.

**Comment by oalders** [comment]
@lolngh ok to close this?

**Comment by lolngh** [comment]
Hi @oalders ,
Could you please provide me some more time?
This issue occurs when perl script is invoked via apache server.
Individually, I didn't find any issue. But ruling out complete possibility of LWP modules with one-more scenario(decrypting the request-response).

**Comment by oalders** [comment]
@lolngh, sure -- let's keep it open.

**Comment by lolngh** [comment]
Hi @noxxi ,
One query - the current perl script is executed via apache server(mod-ssl). So the perl file cannot be ran directly via perl command, are you aware of a way where we can get the debug logs of this perl file when its executed via apache server ?

**Comment by noxxi** [comment]
You can set the debug level also from inside the script

```
use IO::Socket::SSL;
$IO::Socket::SSL::DEBUG=9
```

Note that debugging is done to STDERR. So you either need to look whereever this STDERR ends up in your server or redirect STDERR to a file.

**Comment by lolngh** [comment]
Thanks @noxxi , will try this.

**Comment by lolngh** [comment]
Have tried it, but it wasn't able to save the logs in external file nor in server logs.
For external file, had tried this:
`open( STDERR, ">/tmp/log.txt" );`
`<code>`
`close(STDERR);`

**Comment by lolngh** [comment]
Hi @noxxi ,
Got the below error after going through logs. What might be possible cause and resolution for this?

> DEBUG: .../IO/Socket/SSL.pm:2973: Failed to set SSL cipher list
>DEBUG: .../IO/Socket/SSL.pm:2973: global error: Failed to set SSL cipher list error:1410D0B9:SSL routines:SSL_CTX_set_cipher_list:no cipher match

**Comment by noxxi** [comment]
Some code is trying to set a cipher string which is broken. Might be in your code, might be in the code you use - impossible to say since nothing is known about the code involved. Fix is not to set a broken cipher string. Usually the best is not to set one at all but instead to use the defaults.

**Comment by lolngh** [comment]
Thanks for the information @noxxi , will check about this.

**Comment by lolngh** [comment]
Hi @noxxi ,
Is there a way - where we can enforce the modules to pass in default cipher string ?

**Comment by noxxi** [comment]
> Hi @noxxi , Is there a way - where we can enforce the modules to pass in default cipher string ?

I'm not fully sure what you mean. But if you are faced with some module which sets a broken cipher string and you need to somehow work around this then your can use [set_args_filter_hack](https://metacpan.org/pod/IO::Socket::SSL#IO::Socket::SSL::set_args_filter_hack(\&code|'use_defaults'))

**Comment by lolngh** [comment]
Thanks for the response @noxxi .

**Comment by lolngh** [comment]
Hi @noxxi ,
Sorry for taking more time here.
When I try to run ntlm-auth script indivually in perl file - I don't get any issue.
When apache server via mod-ssl runs this script - that's where we are getting no cipher match error. So are you aware of where the broken cipher might be set ?

Note: Different cipher suite is being used in handshake for both scenarios. ECHDE for apache based run.

**Comment by noxxi** [comment]
No I don't know how your code works inside a different environment. But ciphers configured by LWP have nothing to do with the configuration of the web server where a LWP based script is running. It might be though that execution from a web server opens up a very different environment, maybe using a different SSL configuration, maybe even a different openssl version, maybe even using a different Perl. I don't know anything about your setup though, so all of this is speculation.

> So are you aware of where the broken cipher might be set ?

You can use the [set_args_filter_hack](https://metacpan.org/pod/IO::Socket::SSL#IO::Socket::SSL::set_args_filter_hack(&code%7C'use_defaults')) also to get some stacktrace using Carp::longmess in order to find out where all of this got called from. This might give you some clues on where you need to look.

**Comment by lolngh** [comment]
Thanks for the above suggestion @noxxi. Will try this once.

**Comment by lolngh** [comment]
Hi @noxxi , 
So its is possible to say that - the above issue might have nothing to do with cipher suite setting in NTLM server we are propagating the request to right ?

**Comment by noxxi** [comment]
> Hi @noxxi , So its is possible to say that - the above issue might have nothing to do with cipher suite setting in NTLM server we are propagating the request to right ?

If I understand you correctly this issue happens only when the script is run in the environment of the web application but not when running outside of it. Assuming that the same server is accessed the problem cannot be the server, but it must be the different configuration or software used (openssl version etc) inside the web application environment vs. outside.

**Comment by lolngh** [comment]
The openssl version we are currently using is 1.1.1d. 
Previous openssl version where it was working back then was 1.0.2p. 
When I tried to print out the openssl version(Net::SSLeay::OpenSSL_version_num()) in both the scripts - it give invalid decimal number.
Let me try again.

**Comment by noxxi** [comment]
> The openssl version we are currently using is 1.1.1d. Previous openssl version where it was working back then was 1.0.2p. When I tried to print out the openssl version(Net::SSLeay::OpenSSL_version_num() / Net::SSLeay::SSLeay()) in both the scripts - it give invalid decimal number. Let me try again.

Better use `Net::SSLeay::OPENSSL_VERSION_NUMBER()`. See also https://github.com/noxxi/p5-io-socket-ssl/blob/master/t/01loadmodule.t about the various kind of versions and how they can be retrieved.

**Comment by lolngh** [comment]
Hi @noxxi ,
As you suspected - both scenarios where using different openssl libraries.
Individual - 1.0.2p
Apache - 1.1.1d

Will take a further look into it.

**Comment by lolngh** [comment]
A query, the Authen::NTLM-1.09 is still valid and supported module?

**Comment by noxxi** [comment]
> A query, the Authen::NTLM-1.09 is still valid and supported module? Or should I take up this query to some other team?

Check with the contact given in the module documentation. But a short view at https://metacpan.org/pod/Authen::NTLM shows that the last update was 13 years ago.

**Comment by lolngh** [comment]
Hi @noxxi ,

Thanks for your time & support. We are looking to upgrade the OS(which aligns with packages upgrade).
We will get back if we still face the issue.
